docs/proposals/service-catalog.md

```diff
@@ -0,0 +1,408 @@
+<!-- BEGIN MUNGE: UNVERSIONED_WARNING -->
+
+<!-- BEGIN STRIP_FOR_RELEASE -->
+
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+
+<h2>PLEASE NOTE: This document applies to the HEAD of the source tree</h2>
+
+If you are using a released version of Kubernetes, you should
+refer to the docs that go with that version.
+
+<strong>
+The latest release of this document can be found
+[here](http://releases.k8s.io/release-1.1/docs/proposals/service-catalog.md).
+
+Documentation for other releases can be found at
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+## Abstract
+
+A new catalog concept is proposed for sharing reusable recipes for services,
+the containers that back them, and configuration data associated with them.
+Users will be able to publish recipes to the catalog and browse the catalog
+for recipes to use.
+
+## Motivation
+
+Users don’t like reinventing the wheel. As an example, most users would prefer
```

Comment:
It would be useful to call out the scenarios we want to support more explicitly. I found them difficult to infer from the use cases below.

What's described here is the vanilla whitebox application case: A user wants to instantiate an application from a pre-existing recipe, without forking the recipe. Related cases are forking the recipe first, and creating a new recipe.

There's also the multi-single-tenant case, where the application is instantiated automatically on demand, but is not managed by the consumer -- blackbox to the consumer, but whitebox to the producer. This is probably the most complicated case, and we will want to support it, but maybe not in the initial implementation.

There's the shared-service case: An application is already running, but wants to permit others to consume it. Again, the application is blackbox to the consumer, but whitebox to the producer.

The fully blackbox case is needed for SaaS providers, including cloud providers with hosted applications/middleware, such as RDS and Cloud SQL, and for applications deployed somewhere other than K8s.

A variant of the shared-service case is the private service, where someone just wants to use the binding mechanism to connect a client to their service.